working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,948 @@
+named `box` or `unbox`, said member is not induced.
+
+*The latter rule helps avoiding conflicts in situations where `box` or
+`unbox` is a non-hidden instance member, and it allows developers to write
+their own implementations if needed.*
+
+*The rationale for having this mechanism is that the wrapper object is a
+full-fledged object: It is a subtype of all the direct superinterfaces of
+`E`, so it can be used in code where `E` is not in scope. Moreover, it
+supports late binding of the extension methods, and even dynamic
+invocations. It is costly (it takes space and time to allocate and
+initialize the wrapper object), but it is more robust than the extension
+type, which will only work in a manner which is resolved statically.*
+
+
+## Dynamic Semantics
+
+The dynamic semantics of extension method invocation is very similar to
+the semantics of invocations of the existing extension methods:
+
+If `e` is an expression whose static type is the extension type
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>></code>,
+then a member access like `e.m(args)` is executed by evaluating `e` to an
+object which is bound to a fresh variable `v`, and then evaluating
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>>(v).m(args)</code>,
+and similarly for instance getters and operators.
+
+The dynamic semantics of an invocation of an instance method of the on-type
+which is enabled in an explicit extension type by the show/hide part is as
+if a forwarder were implicitly induced in the extension, with the same
+signature as that of the on-type. *For example:*
+
+```dart
+// Extension type using show/hide:
+extension type MyNum on num show floor {}
+
+// Works like the following:
+extension type MyNum on num {
+  int floor() => this.floor();
+}
+```
+
+*Note that this implies that the extension method `floor` never overrides
+the instance member `floor`, but the extension method `floor` will be
+executed at a call site `myNum.floor()` based on a compile-time decision
+when the receiver `myNum` has static type `MyNum`. In particular, the
+extension method `floor` will never be executed when the receiver has type
+`dynamic`. The forwarding expression `this.floor()` in the implicitly
+induced getter will invoke the instance method, which is subject to late
+binding (so we may end up running `int.floor()` or `double.floor()`,
+depending on the dynamic type of `this`).*
+
+At run time, for a given instance `o` typed as an extension type `E`, there
+is _no_ reification of `E` associated with `o`.
+
+*This means that, at run time, an object never "knows" that it is being
+viewed as having an extension type. By soundness, the run-time type of `o`
+will be a subtype of the on-type of `E`.*
+
+The run-time representation of a type argument which is a non-protected
+extension type `E` (respectively 
+<code>E<T<sub>1</sub>, .. T<sub>k</sub>></code>) 
+is the corresponding instantiated on-type.
+
+*This means that a non-protected extension type and the underlying on-type
```

Comment:
It's pretty weird to me that non-protected are not reified and protected are.

Reply:
Right, it may seem surprising.

The intention is that a non-protected extension type `E` should support maintaining a specific view on objects that were created with no consideration for `E` (say, a JSON value representation built by a third party library using `List<dynamic>` and so on).

In contrast, a protected extension type is optimized for giving developers as much help as possible to maintain the abstraction (the invariants), and then it helps a lot if we can maintain that a `List<MyProtectedInt>` is not simply a `List<int>` at run time, so we don't get to treat it as such.